src/Compilers/Core/Portable/Serialization/ObjectWriter.cs

```diff
@@ -888,15 +890,17 @@ private void WriteObject(object instance, IObjectWritable? instanceAsWritable)
 
                 if (_recursionDepth % MaxRecursionDepth == 0)
                 {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
                     // If we're recursing too deep, move the work to another thread to do so we
-                    // don't blow the stack.  'LongRunning' ensures that we get a dedicated thread
-                    // to do this work.  That way we don't end up blocking the threadpool.
-                    var task = Task.Factory.StartNew(
-                        obj => WriteObjectWorker((IObjectWritable)obj!),
-                        writable,
-                        _cancellationToken,
-                        TaskCreationOptions.LongRunning,
-                        TaskScheduler.Default);
+                    // don't blow the stack.
+                    var task = SerializationThreadPool.RunOnBackgroundThreadAsync(
+                        obj =>
+                        {
+                            WriteObjectWorker((IObjectWritable)obj!);
+                            return null;
+                        },
+                        writable);
```

Comment:
so we return the task, just to synchronously block on the result of it below.  Shoudl we jsut have the helper be a synchronous void method instead and move this blocking into it instead?

Reply:
> Shoudl we jsut have the helper be a synchronous void method instead and move this blocking into it instead?

No. Blocking should always be pushed as far up the stack as possible. This method is the specific method that requires a blocking wait, and therefore this is the only method where the blocking wait is appropriate to appear.

Reply:
I don't agree with that principle. This is a coupled set of functions. But it's not a hill I'd die on here. To me, it provides no value or clarity to have the async concept bleed out to the single location where it is known it will be blocked on (since that's the entire purpose of this component)